### ISSUE TITLE
unexpected behaviour when mixing no_grad decorator with no_grad context manager

### ISSUE THREAD

**Issue by DanielDworakowski** [pytorch/pytorch#10858]
## Issue description

Mixing torch.no_grad decorator with torch.no_grad context re-enables gradients as per torch.is_grad_enabled.

## Code example

```
import torch

@torch.no_grad()
def nothing(x):
    return x

testin = torch.Tensor([0])
with torch.no_grad():
    print(torch.is_grad_enabled())
    testout = nothing(testin)
    print(torch.is_grad_enabled())
```
#### Output:
> False
> True

### Expectation

It would be my expectation that within the no_grad context the gradients would always be disabled.
This expectation matches the output when using nested contexts as below.

```
import torch

def nothing_no_nograd(x):
  return x

testin = torch.Tensor([0])
with torch.no_grad():
    print(torch.is_grad_enabled())
    with torch.no_grad():
        testout = nothing_no_nograd(testin)
    print(torch.is_grad_enabled())
```
#### Output:
> False
> False

## System Info

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 396.54
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] cuda92                    1.0                           0    pytorch
[conda] pytorch                   0.4.1           py36_cuda9.2.148_cudnn7.1.4_1  [cuda92]  pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch
[conda] warpctc-pytorch           0.1                       <pip>
